Initial Linux port

xiangzhai opened the pull request:
Hi,

Build for Linux MIPS64:

```
mkdir build
cd build
cmake .. -DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
-DCORECLR_FOR_CUSTOMGC="/home/loongson/zhaixiang/coreclr-mips64-dev"
make -j8

find . -name "*.so" | xargs file
./src/UpsilonGC/libUpsilonGC.so: ELF 64-bit LSB shared object, MIPS, MIPS64 rel2 version 1 (SYSV), dynamically linked, BuildID[sha1]=78986cee5f86e653a1cdcd9a158c547fcf393968, not stripped
```

CoreCLR Linux MIPS64 port:  https://github.com/gsvm/coreclr

Thanks,
Leslie Zhai

xiangzhai committed:
Initial Linux port

xiangzhai committed:
Initial Linux port

xiangzhai committed:
Initial Linux port

xiangzhai commented:
Hi @kkokosa 

Welcome feedback!

Thanks,
Leslie Zhai

kkokosa commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -500,15 +503,15 @@ bool UpsilonGCHeap::IsInFrozenSegment(Object* object)
 {
     int sizeWithHeader = size + sizeof(ObjHeader);
-    ObjHeader* address = (ObjHeader*)calloc(sizeWithHeader, sizeof(char*));
+    ObjHeader* address = (ObjHeader*)malloc(sizeWithHeader * sizeof(char*));
```
This breaks the allocator contract - we need zero initialized memory returned by `AllocLHeap`.

kkokosa commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -500,15 +503,15 @@ bool UpsilonGCHeap::IsInFrozenSegment(Object* object)
 {
     int sizeWithHeader = size + sizeof(ObjHeader);
-    ObjHeader* address = (ObjHeader*)calloc(sizeWithHeader, sizeof(char*));
+    ObjHeader* address = (ObjHeader*)malloc(sizeWithHeader * sizeof(char*));
```
This breaks the allocator contract - we need zero initialized memory returned by `AllocAllign8`.

kkokosa commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -618,3 +630,24 @@ void UpsilonGCHeap::registerSegment(uint8_t* new_pages)
 {
 	segments[segmentsCount++] = new_pages;
 }
+
+#ifdef __linux__
+uint64_t UpsilonGCHeap::GetTotalAllocatedBytes()
+{
+    return 0;
+}
+
+int UpsilonGCHeap::GetLastGCPercentTimeInGC()
+{
+    return 0;
+}
+
+size_t UpsilonGCHeap::GetLastGCGenerationSize(int gen)
+{
+    return 0;
+}
+
+void UpsilonGCHeap::DiagWalkObject2(Object* obj, walk_fn2 fn, void* context)
+{
+}
+#endif
```
Is this because you need to support newer runtime in your case? If so, I would like to avoid `#ifdef __linux__` and just bump up the supported GC API version by the Upsilon GC (no matter the operating system).

kkokosa commented on src/UpsilonGC/inc/UpsilonGCHandleManager.h:
```diff
@@ -8,9 +11,13 @@ class UpsilonGCHandleManager : public IGCHandleManager
-    virtual void * GetHandleContext(OBJECTHANDLE handle) override;
+    virtual void * GetHandleContext(OBJECTHANDLE handle);
     virtual IGCHandleStore * GetGlobalHandleStore() override;
+#ifdef __linux__
+    virtual IGCHandleStore * CreateHandleStore() override;
+#else
     virtual IGCHandleStore * CreateHandleStore(void * context) override;
+#endif
```
Is this because you need to support newer runtime in your case? If so, I would like to avoid #ifdef __linux__ and just bump up the supported GC API version by the Upsilon GC (no matter the operating system).

kkokosa commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -414,7 +417,7 @@ Object * UpsilonGCHeap::Alloc(gc_alloc_context * acontext, size_t size, uint32_t
 	}
 	int beginGap = 24;
-	uint8_t* newPages = (uint8_t*)VirtualAlloc(NULL, GrowthSize, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
+	uint8_t* newPages = (uint8_t*)VirtualAlloc(nullptr, GrowthSize, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
```
Does it compile? I wonder from where `VirtualAlloc` come from in case of Linux?

xiangzhai replied:
`VirtualAlloc` is available at `coreclr/src/pal/inc/pal.h`

So cmake option is:

```
cmake .. -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DCORECLR_FOR_CUSTOMGC="/home/loongson/zhaixiang/coreclr-mips64-dev"
```

kkokosa reviewed (commented):
Awesome! Thank you for your work. As I see it as an overall the very first port to Linux, I would love to make sure it does not brake anything on Windows and supports the same GC API on Windows and Linux. I left some comments.

xiangzhai committed:
Compiled under Linux MIPS64:

cmake .. -DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
-DCORECLR_FOR_CUSTOMGC="/home/loongson/zhaixiang/coreclr-mips64-dev"

* Zero initialized memory
* Use GC_INTERFACE_MAJOR_VERSION macro
* VirtualAlloc is available at coreclr/src/pal/inc/pal.h

xiangzhai commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -500,15 +502,17 @@ bool UpsilonGCHeap::IsInFrozenSegment(Object* object)
     int sizeWithHeader = size + sizeof(ObjHeader);
-    ObjHeader* address = (ObjHeader*)calloc(sizeWithHeader, sizeof(char*));
+    ObjHeader* address = (ObjHeader*)malloc(sizeWithHeader * sizeof(char*));
+    memset(address, 0, sizeWithHeader * sizeof(char*));
```
Thanks for pointing out my fault!

xiangzhai commented on src/UpsilonGC/UpsilonGCHeap.cpp:
```diff
@@ -601,7 +605,16 @@ void UpsilonGCHeap::ControlPrivateEvents(GCEventKeyword keyword, GCEventLevel le
 {
 }
 
+#if GC_INTERFACE_MAJOR_VERSION >= 3
```
Thanks for your advice!
